File: CivitasJava/src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 * @author Miguel Muñoz Molina
 * @author Iván Valero Rodríguez
 */

public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas ;
    private boolean barajada;
    private int usadas ;
    private boolean debug ;
    private ArrayList<Sorpresa> cartasEspeciales ;
    private Sorpresa ultimaSorpresa ;
    private Diario diario = Diario.getInstance() ;
    
    private void init(){
        sorpresas = new ArrayList<Sorpresa>() ;
        cartasEspeciales = new ArrayList<Sorpresa>() ;
        barajada = false ;
        usadas = 0 ;
    }
    
    MazoSorpresas (boolean modo){
        init();
        debug = modo ;
        if (debug){
            diario.ocurreEvento("Modo debug para el mazo: Activado");
        }
    }
        
    MazoSorpresas (){
        init();
        debug = false ;
    }
    
    void alMazo(Sorpresa s){
        if (!barajada)
            sorpresas.add(s);
    }
    
    Sorpresa siguiente(){
        if (!barajada || usadas == sorpresas.size()){
            Collections.shuffle(sorpresas);
            usadas = 0;
            barajada = true ;
        }
        usadas++ ;
        ultimaSorpresa = sorpresas.get(0) ;
        sorpresas.remove(0) ;
        sorpresas.add(ultimaSorpresa) ;
        return ultimaSorpresa ;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        if(sorpresas.contains(sorpresa)){
            sorpresas.remove(sorpresa) ;
            cartasEspeciales.add(sorpresa) ;
            diario.ocurreEvento("Se ha inhabilitado la sorpresa :" + sorpresa.toString());
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        if(cartasEspeciales.contains(sorpresa)){
            cartasEspeciales.remove(sorpresa) ;
            sorpresas.add(sorpresa) ;
            diario.ocurreEvento("Se ha habilitado la sorpresa :" + sorpresa.toString());
        }
    }
    
}

File: CivitasRuby/lib/MazoSorpresas.rb
# encoding:utf-8

#Author: Miguel Muñoz Molina
#Author: Iván Valero Rodríguez

require_relative 'diario'
require_relative "Sorpresa"
module Civitas
  class MazoSorpresas 
    def init
      @sorpresas = Array.new
      @cartasEspeciales = Array.new 
      @barajada = false
      @usadas = 0
      @diario = Diario.instance
      @debug
      @ultimaSorpresa
    end
    
    def initialize(modo = false)
      init
      @debug = modo
      if(@debug)
        @diario.ocurre_evento("Modo debug para el mazo: Activado")
      end
    end
    
    def alMazo(s)
      if(!@barajada)
        @sorpresas << s
      end
    end
    
    def siguiente
      if (!@barajada || @usadas == @sorpresas.size)
        @sorpresas.shuffle
        @usadas = 0
        @barajada = true
      end
      @usadas = @usadas + 1
      @ultimaSorpresa = @sorpresas.at(0)
      @sorpresas.delete_at(0)
      @sorpresas << @ultimaSorpresa
      return @ultimaSorpresa
    end
    
    def inhabilitarCartaEspecial(sorpresa)
      if(@sorpresas.include?(sorpresa))
        @sorpresas.delete_at(@sorpresas.index(sorpresa))
        @cartasEspeciales << sorpresa
        @diario.ocurre_evento("Se ha inhabilitado la sorpresa #{sorpresa.to_s}")
      end
    end
    
    def habilitarCartaEspecial(sorpresa)
      if(@cartasEspeciales.include?(sorpresa))
        @cartasEspeciales.delete_at(@sorpresas.index(sorpresa).to_i)
        @sorpresas << sorpresa
        @diario.ocurre_evento("Se ha habilitado la sorpresa #{sorpresa.to_s}")
      end
    end
    
    private :init
    
  end
end
